src/backend/mpi/write_buffer.hpp

```diff
@@ -0,0 +1,307 @@
+			return _buffer.at(i);
+		}
+
+		/**
+		 * @brief	Check if val exists in the buffer
+		 * @param	val The value to check for
+		 * @return	True if val exists in the buffer, else False
+		 */
+		bool has(T val) {
+			typename std::deque<T>::iterator it = std::find(_buffer.begin(),
+					_buffer.end(), val);
+			return (it != _buffer.end());
+		}
+
+		/**
+		 * @brief	Constructs a new element and emplaces it at the back of the buffer
+		 * @param	args Properties of the new element to emplace back
+		 */
+		template<typename... Args>
+			void emplace_back( Args&&... args) {
+				_buffer.emplace_back(std::forward<Args>(args)...);
+			}
+
+		/**
+		 * @brief	Removes the front element from the buffer and returns it
+		 * @return	The element that was popped from the buffer
+		 */
+		T pop() {
+			auto elem = std::move(_buffer.front());
+			_buffer.pop_front();
+			return elem;
+		}
+
+		/**
+		 * @brief	Sorts all elements in ascending order
+		 */
+		void sort() {
+			std::sort(_buffer.begin(), _buffer.end());
+		}
+
+		/**
+		 * @brief	Sorts the first _write_back_size elements in ascending order
+		 */
+		void sort_first() {
+			assert(_buffer.size() >= _write_back_size);
+			std::sort(_buffer.begin(), _buffer.begin()+_write_back_size);
+		}
+
+		/**
+		 * @brief	Flushes first _write_back_size elements of the  ArgoDSM 
+		 * 			write buffer to memory
+		 * @pre		Require ibsem to be taken until parallel MPI
+		 * @pre		Require write_buffer_mutex to be held
+		 */
+		void flush_partial() {
+			// Sort the first _write_back_size elements
+			sort_first();
+
+			// For each element, handle the corresponding ArgoDSM page
+			for(std::size_t i = 0; i < _write_back_size; i++) {
+				// The code below should be replaced with a cache API
+				// call to write back a cached page
+				std::size_t cache_index = pop();
+				std::uintptr_t page_address = cacheControl[cache_index].tag;
+				void* page_ptr = static_cast<char*>(
+						argo::virtual_memory::start_address()) + page_address;
+
+				// Write back the page
+				mprotect(page_ptr, block_size, PROT_READ);
+				cacheControl[cache_index].dirty=CLEAN;
+				for(int i=0; i < CACHELINE; i++){
+					storepageDIFF(cache_index+i,page_size*i+page_address);
+				}
+			}
+
+			// Close any windows used to write back data
+			// This should be replaced with an API call
+			for(int i = 0; i < argo::backend::number_of_nodes(); i++){
+				if(barwindowsused[i] == 1){
+					MPI_Win_unlock(i, globalDataWindow[i]);
+					barwindowsused[i] = 0;
+				}
+			}
+		}
+
+	public:
+		/**
+		 * @brief	Constructor
+		 */
+		write_buffer()	:
+			_max_size(argo::env::write_buffer_size()/CACHELINE),
+			_write_back_size(argo::env::write_buffer_write_back_size()/CACHELINE) { }
+
+		/**
+		 * @brief	Copy constructor
+		 * @param	other The write_buffer object to copy from
+		 */
+		write_buffer(const write_buffer & other) {
+			// Ensure protection of data
+			std::unique_lock<std::mutex> lock_other(other._buffer_mutex);
```

Comment:
I was debating this when I wrote the code, and seeing as you also hold this opinion I changed the locks that do not rely on deferred locking to use `std::lock_guard`. In c++17 `std::lock_guard` is superceded by `std::scoped_lock`, just as a note in case we would migrate at some point.